test/lit/passes/dae-gc.wast

```diff
@@ -594,4 +600,104 @@
  (func $return-ref-func (result funcref)
   (ref.func $do-return-call)
  )
+
+ ;; Show that we can optimize the return type of a function that does a tail
+ ;; call.
+ ;; CHECK:      (func $tail-callee (result (ref ${}))
+ ;; CHECK-NEXT:  (unreachable)
+ ;; CHECK-NEXT: )
+ (func $tail-callee (result (ref ${}))
+  (unreachable)
+ )
+ ;; CHECK:      (func $tail-caller-yes (result (ref ${}))
+ ;; CHECK-NEXT:  (return_call $tail-callee)
+ ;; CHECK-NEXT: )
+ (func $tail-caller-yes (result anyref)
+  ;; This function's return type can be refined because of this call, whose
+  ;; target's return type is more specific than anyref.
+  (return_call $tail-callee)
+ )
+ ;; CHECK:      (func $tail-caller-no (result anyref)
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (i32.const 1)
+ ;; CHECK-NEXT:   (return
+ ;; CHECK-NEXT:    (ref.null any)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (return_call $tail-callee)
+ ;; CHECK-NEXT: )
+ (func $tail-caller-no (result anyref)
+  ;; This function's return type cannot be refined because of another return
+  ;; whose type prevents it.
+  (if (i32.const 1)
+   (return (ref.null any))
+  )
+  (return_call $tail-callee)
+ )
```

Comment:
For these *-no tests, it would be more interesting if the tail call type blocked the optimization, since we know from previous tests that normal returns can block the optimization.

Reply:
Hmm, perhaps, but the test it is most directly compared to is the one right above it, so the smaller the diff between them the better I think.

I can add an additional flipped test, if you think that's worth it? As the logic for all returns is identical (they all call the same shared method) that feels a little unnecessary to me personally.

Reply:
Fair enough!